Check in a minimal JAX custom call example using XLA FFI

andportnoy opened the pull request.

andportnoy committed:
Check in a minimal JAX custom call example using XLA FFI

andportnoy committed:
Remove pybind11 dependency, build PyCapsule in Python

andportnoy committed:
Use XLA FFI headers that are shipped with JAX

dfm commented on docs/cuda-custom-call/mul.cu:
```diff
@@ -0,0 +1,42 @@
+namespace ffi = xla::ffi;
+
+// XLA FFI binding wrapper around our host launcher function
+extern "C" XLA_FFI_Error *Mul(XLA_FFI_CallFrame *call_frame) {
+  static const auto *kImpl =
+      ffi::Ffi::Bind()
+          .Ctx<ffi::PlatformStream<cudaStream_t>>()
+          .Arg<ffi::Buffer<ffi::DataType::F32>>()
+          .Arg<ffi::Buffer<ffi::DataType::F32>>()
+          .Ret<ffi::Buffer<ffi::DataType::F32>>()
+          .Attr<size_t>("n")
+          .To([](cudaStream_t stream,
+                 ffi::Buffer<ffi::DataType::F32> a,
+                 ffi::Buffer<ffi::DataType::F32> b,
+                 ffi::Result<ffi::Buffer<ffi::DataType::F32>> c,
+                 size_t n) -> ffi::Error {
+            LaunchMulKernel(stream, a.data, b.data, c->data, n);
+            return ffi::Error::Success();
+      }).release();
+  return kImpl->Call(call_frame);
+}
```
Not crucial, but some of this boilerplate could be abstracted with the [`XLA_FFI_DEFINE_HANDLER` macro](https://github.com/openxla/xla/blob/6d17fd8fbe27b06fc4828dc0c38ecdd8f9ba8573/xla/ffi/api/api.h#L1602). This is the syntax used in the [XLA custom call docs](https://github.com/openxla/xla/blob/6d17fd8fbe27b06fc4828dc0c38ecdd8f9ba8573/docs/custom_call.md) now.

andportnoy replied:
Thanks, this looks like this should reduce the amount of boilerplate. The flip side is I'm now looking at a template error that I can't understand :)

andportnoy replied:
The error was due to the fact that the function I was exporting using this macro has return type `void`, but it needs to be an `ffi::Error` instead.

But now I'm facing a different issue: the macro creates a local lambda variable instead of a function definition. This doesn't work for my case because I'm trying to export a function with C linkage.

andportnoy replied:
One solution suggested by @ezhulenev is to add a new macro that defines a function with C linkage.

ezhulenev replied:
Yeah, basically `XLA_FFI_EXPORT_HANDLER` that will hide the `extern "C" XLA_FFI_Error *Mul` boilerplate. Another options is to change default behavior of `XLA_FFI_DEFINE_HANDLER` to define C function by default and add another one for users that prefer lambda.

andportnoy committed:
Fix build_capsule to use its argument

andportnoy committed:
Reflow comment above build_capsule

andportnoy committed:
Replace string literals with variables

andportnoy committed:
Use c = a * (b+1) instead of elementwise multiplication

andportnoy committed:
Check for CUDA errors in XLA FFI wrapper

andportnoy committed:
Implement custom_vjp for foo

andportnoy committed:
Fix comment about foo_bwd

andportnoy committed:
Add pycapsule utility to jax.ffi

This allows one to build PyCapsules ready for XLA custom call
registration out of function pointers retrieved from shared libraries
using ctypes. In particular, this obviates the need to create Python
bindings to custom call targets.

andportnoy committed:
Move CUDA custom call example to tests directory

andportnoy committed:
Add BUILD file for cuda_custom_call, rename foo.cu -> foo.cu.cc

andportnoy committed:
Add pycapsule utility to jax.extend.ffi

This allows one to build PyCapsules ready for XLA custom call
registration out of function pointers retrieved from shared libraries
using ctypes. In particular, this obviates the need to create Python
bindings to custom call targets.

andportnoy committed:
Move build_capsule to jax.ffi

andportnoy committed:
Start integrating the example as a Bazel test

andportnoy commented:
Going to keep pushing to this PR for now, but will open a new one when it's ready.

andportnoy committed:
Specify jax.extend as an explicit test dependency

Otherwise this subpackage is not available at test runtime.

andportnoy committed:
Remove JAX version print

andportnoy committed:
Fix test time path to libfoo.so

andportnoy committed:
Get the cuda_custom_call test to work

andportnoy committed:
Remove conversion to tuple in foo_bwd

Fixed by https://github.com/google/jax/pull/21568.

andportnoy committed:
Improve comments in cuda_custom_call_test

andportnoy committed:
Remove test debug code

andportnoy committed:
Turn cuda_custom_call_test into a JAX test

andportnoy committed:
Format using yapf

yapf -i --style='{based_on_style: google, indent_width: 2}' ...
